server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -77,6 +77,7 @@
 public class IndexShardRoutingTable implements Iterable<ShardRouting> {
 
     final ShardShuffler shuffler;
+    final ShardShuffler shufflerForWeightedRouting;
```

Comment:
Do we need a second shuffler or just creating another method the prevents the shuffle for `onlyNodeSelectorActiveInitializingShardsIt` should work?

Reply:
The issue is with this line  https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java#L587. This is moving the shuffler one more time ie in a three node cluster with weights (1,1,0) the second request also hits the same node since the shuffler moves by 2 in the first request.

Since we need `onlyNodeSelectorActiveInitializingShardsIt` to get shards belonging to node with weight zero. Introducing a method without shuffler will require a code duplication.

Reply:
Lets put up a comment explaining the same.